File: samples/macOS/cs/netcore1.0/NonTableManagerSamples/ShowCommissionDetails/src/SessionStatusListener.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using fxcore2;
using ArgParser;

namespace ShowCommissionDetails
{
    class SessionStatusListener : IO2GSessionStatus
    {
        private bool mConnected;
        private bool mDisconnected;
        private bool mError;
        private O2GSession mSession;
        private EventWaitHandle mSyncSessionEvent;
        private LoginParams mLoginParams;

        /// <summary>
        /// ctor
        /// </summary>
        /// <param name="session"></param>
        public SessionStatusListener(O2GSession session, LoginParams loginParams)
        {
            mSession = session;
            mLoginParams = loginParams;
            Reset();
            mSyncSessionEvent = new EventWaitHandle(false, EventResetMode.AutoReset);
        }

        public bool Connected
        {
            get
            {
                return mConnected;
            }
        }

        public bool Disconnected
        {
            get
            {
                return mDisconnected;
            }
        }

        public bool Error
        {
            get
            {
                return mError;
            }
        }

        public void Reset()
        {
            mConnected = false;
            mDisconnected = false;
            mError = false;
        }

        public bool WaitEvents()
        {
            return mSyncSessionEvent.WaitOne(30000);
        }

        public void onSessionStatusChanged(O2GSessionStatusCode status)
        {
            Console.WriteLine("Status: " + status.ToString());
            switch (status)
            {
                case O2GSessionStatusCode.TradingSessionRequested:
                    O2GSessionDescriptorCollection descs = mSession.getTradingSessionDescriptors();
                    Console.WriteLine("Session descriptors:");
                    Console.WriteLine("id, name, description, requires pin");
                    foreach (O2GSessionDescriptor desc in descs)
                    {
                        Console.WriteLine("'{0}' '{1}' '{2}' {3}", desc.Id, desc.Name, desc.Description, desc.RequiresPin);
                    }
                    Console.WriteLine();

                    string sSessionID;
                    if (string.IsNullOrEmpty(mLoginParams.SessionID))
                    {
                        Console.WriteLine("Please enter trading session ID and press \'Enter\'");
                        sSessionID = Console.ReadLine();
                    }
                    else
                    {
                        sSessionID = mLoginParams.SessionID;
                    }

                    mSession.setTradingSession(sSessionID.Trim(), mLoginParams.Pin);
                    break;
                case O2GSessionStatusCode.Connected:
                    mConnected = true;
                    mDisconnected = false;
                    mSyncSessionEvent.Set();
                    break;
                case O2GSessionStatusCode.Disconnected:
                    mConnected = false;
                    mDisconnected = true;
                    mSyncSessionEvent.Set();
                    break;
            }
        }

        public void onLoginFailed(string error)
        {
            Console.WriteLine("Login error: " + error);
            mError = true;
        }
    }
}

File: samples/macOS/java/TableManagerSamples/CalculateTradingCommissions/src/SessionStatusListener.java
package calculatetradingcommissions;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;

import com.fxcore2.*;

public class SessionStatusListener implements IO2GSessionStatus {
    private O2GSession mSession;
    private boolean mConnected;
    private boolean mDisconnected;
    private boolean mError;
    private Semaphore mSemaphore;

    // ctor
    public SessionStatusListener(O2GSession session) {
        mSession = session;
        reset();
        mSemaphore = new Semaphore(0);
    }

    public boolean isConnected() {
        return mConnected;
    }

    public boolean isDisconnected() {
        return mDisconnected;
    }

    public boolean hasError() {
        return mError;
    }

    public void reset() {
        mConnected = false;
        mDisconnected = false;
        mError = false;
    }

    public boolean waitEvents() throws Exception {
        return mSemaphore.tryAcquire(30, TimeUnit.SECONDS);
    }

    public void onSessionStatusChanged(O2GSessionStatusCode status) {
        System.out.println("Status: " + status.toString());
        switch (status) {
        case TRADING_SESSION_REQUESTED:
            O2GSessionDescriptorCollection descs = mSession.getTradingSessionDescriptors();
            System.out.println("Session descriptors:");
            System.out.println("id, name, description, requires pin");
            for (O2GSessionDescriptor desc : descs) {
                System.out.println(String.format("'%s' '%s' '%s' %s", desc.getId(), desc.getName(), desc.getDescription(), desc.isPinRequired()));
            }
            
            InputStreamReader converter = new InputStreamReader(System.in);
            BufferedReader in = new BufferedReader(converter);
            System.out.println("Please enter trading session ID and press \'Enter\'");
            String sSessionID = "";
            try {
                sSessionID = in.readLine();
            } catch (IOException e) {
            }
            System.out.println("Please enter pin (if required). Then press \'Enter\'");
            String sPin = "";
            try {
                sPin = in.readLine();
            } catch (IOException e) {
            }

            mSession.setTradingSession(sSessionID.trim(), sPin.trim());
            break;
        case CONNECTED:
            mConnected = true;
            mDisconnected = false;
            mSemaphore.release();
            break;
        case DISCONNECTED:
            mConnected = false;
            mDisconnected = true;
            mSemaphore.release();
            break;
        }
    }

    public void onLoginFailed(String sError) {
        System.out.println("Login error: " + sError);
        mError = true;
    }
}
